<a href="https://colab.research.google.com/github/shamim13u/Solar_Panel_Analysis/blob/main/solar_panel_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> Data Analytics of Photovoltaic System
</center></h2>



<center>
<img src="https://github.com/DorobantuDiana/DS_Projects/blob/main/objects/solar_power.gif?raw=true" width=850>
</center>

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> Data Analytics of Photovoltaic System
</center></h2>  
    
* <b>Solar Panels</b>, also known as <i>photovoltaic (PV) panels</i>, are devices that <u>convert Sunlight into Electricity</u> using the photovoltaic effect; they play a crucial role in the <i>renewable energy sector</i>, providing a clean and sustainable source of power; solar panels have gained significant popularity due to their environmental benefits, potential cost savings and increasing energy demands.

* by leveraging <b>data Analytics</b> techniques, solar panel projects can benefit from improved <b><i>energy forecasting</i></b>, optimized <b><i>system performance</i></b>, early <b><i>fault detection</i></b> and efficient <b><i>energy distribution</i></b>; these data-driven insights can enhance the solar energy system's:

<ol style="margin-left:360px">
    <li><b>overall efficiency</b> </li>
    <li><b>reliability</b></li>
    <li><b>cost-effectiveness</b> </li>
</ol>

---
<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> TABLE OF CONTENTS
</center></h2>   

1. [Dataset](#1)
    
2. [Libraries](#2)
   
3. [Data Loading](#3)

4. [Describe the data](#4)

5. [Check if any missing values](#5)

6. [Total Number of Invertor](#6)

7. [EDA](#7)
    
8. [System Performance Analysis](#8)

9. [Performance Ratio](#9)


     
<br>  
    
---

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> LIBRARIES
</center></h2>

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> DATA LOADING
</center></h2>

In [3]:
# power generation datasets
Plant_1_Generation_Data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/sp/Plant_1_Generation_Data.csv")
Plant_2_Generation_Data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/sp/Plant_2_Generation_Data.csv")
# sensor datasets
Plant_1_weather_sensor_Data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/sp/Plant_1_Weather_Sensor_Data.csv")
Plant_2_weather_sensor_Data = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/sp/Plant_2_Weather_Sensor_Data.csv")




In [ ]:
Plant_1_Generation_Data.head()

In [ ]:
Plant_2_Generation_Data.head()

In [ ]:
Plant_1_weather_sensor_Data.head()

In [ ]:
Plant_2_weather_sensor_Data.head()

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> DESCRIBE THE DATA
</center></h2>


In [ ]:
Plant_1_Generation_Data.describe()

In [ ]:
Plant_2_Generation_Data.describe()


In [ ]:
Plant_1_weather_sensor_Data.describe()

In [ ]:
Plant_2_weather_sensor_Data.describe()

In [ ]:
Plant_1_Generation_Data.head()

In [14]:
Plant_2_Generation_Data.head()

DATE_TIME  PLANT_ID       SOURCE_KEY  DC_POWER  AC_POWER  \
0  2020-05-15 00:00:00   4136001  4UPUqMRk7TRMgml       0.0       0.0   
1  2020-05-15 00:00:00   4136001  81aHJ1q11NBPMrL       0.0       0.0   
2  2020-05-15 00:00:00   4136001  9kRcWv60rDACzjR       0.0       0.0   
3  2020-05-15 00:00:00   4136001  Et9kgGMDl729KT4       0.0       0.0   
4  2020-05-15 00:00:00   4136001  IQ2d7wF4YD8zU1Q       0.0       0.0   

   DAILY_YIELD   TOTAL_YIELD  
0  9425.000000  2.429011e+06  
1     0.000000  1.215279e+09  
2  3075.333333  2.247720e+09  
3   269.933333  1.704250e+06  
4  3177.000000  1.994153e+07

In [ ]:
Plant_1_weather_sensor_Data.head()

In [ ]:
Plant_2_weather_sensor_Data

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> CHECK MISSING VALUES
</center></h2>

In [16]:
Plant_1_Generation_Data.isnull().sum()

DATE_TIME      0
PLANT_ID       0
SOURCE_KEY     0
DC_POWER       0
AC_POWER       0
DAILY_YIELD    0
TOTAL_YIELD    0
dtype: int64

In [ ]:
Plant_2_Generation_Data.isnull().sum()

In [ ]:
Plant_1_weather_sensor_Data.isnull().sum()

In [ ]:
Plant_2_weather_sensor_Data.isnull().sum()

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> MAXIMUM AMBIENT TEMPERATURE
</center></h2>

In [ ]:
Plant_1_ambient_temperature = Plant_1_weather_sensor_Data['AMBIENT_TEMPERATURE']
max(round(Plant_1_ambient_temperature,2))

In [ ]:
Plant_2_ambient_temperature = Plant_2_weather_sensor_Data['AMBIENT_TEMPERATURE']
max(round(Plant_2_ambient_temperature,2))

In [ ]:
Plant_1_Generation_Data.iloc[[max(Plant_1_Generation_Data['DC_POWER'])]]

In [ ]:
Plant_2_Generation_Data.iloc[[max(Plant_2_Generation_Data['DC_POWER'])]]

In [ ]:
Plant_1_Generation_Data.sort_values("DC_POWER", axis = 0, ascending = False,
                 inplace = True, na_position ='last')
Plant_1_Generation_Data

In [ ]:
Plant_2_Generation_Data.sort_values("DC_POWER", axis = 0, ascending = False,
                 inplace = True, na_position ='last')
Plant_2_Generation_Data

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
     </br> TOTAL NUMBER OF INVERTOR

</center></h2>

In [ ]:
print("The number of inverters in plant_1 = ",Plant_1_Generation_Data ['SOURCE_KEY'].nunique())

In [ ]:
print("The number of inverters in plant_2 = ",Plant_2_Generation_Data ['SOURCE_KEY'].nunique())

<div style="border-radius: 40px;
            color: #403b34;
            background-color: #bdbebf;
            font-size: 100%;
            text-align: left;
            padding: 20px;
            margin: 10px;
            box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <p style="margin-bottom: 10px; font-weight: bold;">Output Analysis:</p>
    <ul style="margin-top: 0;">
        <li>Both Plant-1 & Plant-2 have 22 Invertor each
    </ul>
</div>

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> EXPLORATORY DATA ANALYSIS (EDA)
</center></h2>

In [ ]:
# power generation datasets
def check(data):
    l = []
    columns = data.columns
    for col in columns:
        instances = data[col].count()
        dtypes = data[col].dtype
        unique = data[col].nunique()
        sum_null = data[col].isnull().sum()
        duplicates = data.duplicated().sum()
        l.append([col, dtypes, instances, unique, sum_null, duplicates])
    data_check = pd.DataFrame(l, columns=["column", "dtype", "instances", "unique", "sum_null", "duplicates"])
    return data_check

check1 = check(Plant_1_Generation_Data )
check2 = check(Plant_2_Generation_Data )

output = pd.concat([check1, check2], axis=1)

output


In [ ]:
Plant_1_Generation_Data.info()

In [ ]:
Plant_2_Generation_Data.info()

In [ ]:
print("\n")
Plant_1_Generation_Data.describe()

In [ ]:
Plant_2_Generation_Data.info()

In [ ]:
print ("\n")
Plant_2_Generation_Data.describe()

<div style="border-radius: 40px;
            color: #403b34;
            background-color: #bdbebf;
            font-size: 100%;
            text-align: left;
            padding: 20px;
            margin: 10px;
            box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <p style="margin-bottom: 10px; font-weight: bold;">Output Analysis:</p>
    <ul style="margin-top: 0;">
        <li>power generation dataset of Plant 1 comprises a total of 68,778 records</li>
        <li>power generation dataset of Plant 2 comprises a total of 67,698 records
        <li>there are neither duplicates nor any missing values in both power generation datasets</li>
        <li>each of the two plants is equipped with a total of 22 inverters</li>
        <li>'DATE_TIME' column is currently stored as an object data type instead of date-time in both power generation datasets</li>
    </ul>
</div>

In [ ]:
# sensor datasets
def check(data):
    l = []
    columns = data.columns
    for col in columns:
        instances = data[col].count()
        dtypes = data[col].dtype
        unique = data[col].nunique()
        sum_null = data[col].isnull().sum()
        duplicates = data.duplicated().sum()
        l.append([col, dtypes, instances, unique, sum_null, duplicates])
    data_check = pd.DataFrame(l, columns=["column", "dtype", "instances", "unique", "sum_null", "duplicates"])
    return data_check

check1 = check(Plant_1_weather_sensor_Data)
check2 = check(Plant_1_weather_sensor_Data)

output = pd.concat([check1, check2], axis=1)

output

In [ ]:
Plant_1_weather_sensor_Data.info()

In [35]:
print("\n")
Plant_1_weather_sensor_Data.describe()

PLANT_ID  AMBIENT_TEMPERATURE  MODULE_TEMPERATURE  IRRADIATION
count     3182.0          3182.000000         3182.000000  3182.000000
mean   4135001.0            25.531606           31.091015     0.228313
std          0.0             3.354856           12.261222     0.300836
min    4135001.0            20.398505           18.140415     0.000000
25%    4135001.0            22.705182           21.090553     0.000000
50%    4135001.0            24.613814           24.618060     0.024653
75%    4135001.0            27.920532           41.307840     0.449588
max    4135001.0            35.252486           65.545714     1.221652

In [ ]:
Plant_2_weather_sensor_Data.info()

In [ ]:
print("\n")
Plant_2_weather_sensor_Data.describe()

In [ ]:
Plant_1_weather_sensor_Data.head()

<div style="border-radius: 40px;
            color: #403b34;
            background-color: #bdbebf;
            font-size: 100%;
            text-align: left;
            padding: 20px;
            margin: 10px;
            box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <p style="margin-bottom: 10px; font-weight: bold;">Output Analysis:</p>
    <ul style="margin-top: 0;">
        <li>sensor dataset of Plant 1 comprises a total of 3,182 records</li>
        <li>sensor generation dataset of Plant 2 comprises a total of 3,259 records
        <li>there are neither duplicates nor any missing values in both  sensor datasets</li>
        <li>'DATE_TIME' column is currently stored as an object data type instead of date-time in both sensor datasets</li>
    </ul>
</div>

In [39]:
# Parse datetime
Plant_1_Generation_Data["DATE_TIME"] = pd.to_datetime(Plant_1_Generation_Data["DATE_TIME"],format="%d-%m-%Y %H:%M")
Plant_2_Generation_Data["DATE_TIME"] = pd.to_datetime(Plant_2_Generation_Data["DATE_TIME"],format="%Y-%m-%d %H:%M:%S")
Plant_1_weather_sensor_Data["DATE_TIME"] = pd.to_datetime(Plant_1_weather_sensor_Data["DATE_TIME"],format="%Y-%m-%d %H:%M:%S")
Plant_2_weather_sensor_Data["DATE_TIME"] = pd.to_datetime(Plant_2_weather_sensor_Data["DATE_TIME"],format="%Y-%m-%d %H:%M:%S")


Plant_1_Generation_Data_idx = Plant_1_Generation_Data.groupby('DATE_TIME').sum().reset_index()
Plant_1_Generation_Data_idx["PLANT_ID"]= 1
Plant_2_Generation_Data_idx = Plant_2_Generation_Data.groupby('DATE_TIME').sum().reset_index()
Plant_2_Generation_Data_idx["PLANT_ID"]= 2

Plant_1_weather_sensor_Data["PLANT_ID"]= 1
Plant_2_weather_sensor_Data["PLANT_ID"]= 2


In [ ]:
Plant_2_Generation_Data.head()

<div style="border-radius: 40px;
            color: #403b34;
            background-color: #bdbebf;
            font-size: 100%;
            text-align: left;
            padding: 20px;
            margin: 10px;
            box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <p style="margin-bottom: 10px; font-weight: bold;">Output Analysis:</p>
    <ul style="margin-top: 0;">
        <li>In the power generation dataset of Plant 1, there seems to be a big difference between the AC power and DC power measurements. This difference is so significant that it suggests the possibility of a tenfold (10 times) gap between the recorded values for DC power. </li>
        </ul>
</div>

In [ ]:
fig, axs = plt.subplots(1, 2, figsize=(20, 5))


axs[0].plot(Plant_2_Generation_Data_idx['DATE_TIME'], Plant_2_Generation_Data_idx['DC_POWER'], color='#cc7e08')
axs[0].set_xlabel('Date')
axs[0].set_ylabel('kW')
axs[0].set_title('DC Power - Plant 2')
axs[0].tick_params(axis='x', rotation=45)


axs[1].plot(Plant_2_Generation_Data_idx['DATE_TIME'], Plant_2_Generation_Data_idx['AC_POWER'], color='#2b2924')
axs[1].set_xlabel('Date')
axs[1].set_ylabel('kW')
axs[1].set_title('AC Power - Plant 2')
axs[1].tick_params(axis='x', rotation=45)


plt.tight_layout()
plt.show()

plt.figure(figsize=(20, 5))
plt.plot(Plant_2_Generation_Data_idx['DATE_TIME'], Plant_2_Generation_Data_idx['DC_POWER'], color='#cc7e08', label='DC Power')
plt.plot(Plant_2_Generation_Data_idx['DATE_TIME'], Plant_2_Generation_Data_idx['AC_POWER'], color='#2b2924', label='AC Power')
plt.xlabel('Date')
plt.ylabel('kW')
plt.title('Power Output - Plant 2')
plt.xticks(rotation=45)
plt.legend()
plt.tight_layout()
plt.show()


<div style="border-radius: 40px;
            color: #403b34;
            background-color: #bdbebf;
            font-size: 100%;
            text-align: left;
            padding: 20px;
            margin: 10px;
            box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <p style="margin-bottom: 10px; font-weight: bold;">Output Analysis:</p>
    <ul style="margin-top: 0;">
        <li> AC power and DC power measurements are relatively close or consistent with each other.  </li>
        </ul>
</div>

In [42]:
Plant_1_Generation_Data_idx["DC_POWER"] = Plant_1_Generation_Data_idx["DC_POWER"]/10

# Merging Generation Data
power_ds = pd.concat([Plant_1_Generation_Data_idx, Plant_2_Generation_Data_idx])

# Merging Sensor Data
sensor_ds = pd.concat([Plant_1_weather_sensor_Data, Plant_2_weather_sensor_Data])

ds = pd.merge(power_ds, sensor_ds, on=['DATE_TIME','PLANT_ID'])


In [ ]:
# plot Daiyly Yield against plants
plt.figure(figsize=(20, 5))
sns.lineplot(x="DATE_TIME", y="DAILY_YIELD",
             hue="PLANT_ID",
             palette=["#cc7e08", "#2b2924"],
             data=ds)
plt.tight_layout()

In [ ]:
correlation_matrix = ds.corr()

# Set up the figure and axes
fig, ax = plt.subplots(figsize=(8, 6))

# Create the correlogram using seaborn's pairplot function
sns.heatmap(correlation_matrix, annot=True, fmt=".2f", cmap="copper", ax=ax)

# Customize the plot aesthetics
ax.set_title("Correlation Plot", fontsize=14)
ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right', fontsize=8)
ax.set_yticklabels(ax.get_yticklabels(), rotation=0, horizontalalignment='right', fontsize=8)

# Show the plot
plt.tight_layout()
plt.show()


<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:1; border-radius: 15px; font-weight: bold; color:#your_text_color'><center>
    </br>SYSTEM PERFORMANCE
</center></h2>

<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> Conversion Efficiency
</center></h2>  

<ol style="margin-left:10px;margin-left:10px">
    
It provides an indication of how effectively the panels are converting sunlight into usable electricity. This can be calculated using the following formula:
</ol>

<ol style="margin-left:360px">
    <b>Conversion Efficiency = (AC Power / DC Power) * 100</b>
</ol>

In [45]:
ConvEff_Plant1 = (sum(Plant_1_Generation_Data_idx['AC_POWER']) / (sum(Plant_1_Generation_Data_idx['DC_POWER']))) * 100

print("The Conversion Efficiency for Plant 1 is", round(ConvEff_Plant1, 3), "%")

The Conversion Efficiency for Plant 1 is 97.795 %


In [46]:
ConvEff_Plant2 = (sum(Plant_2_Generation_Data_idx['AC_POWER']) / sum(Plant_2_Generation_Data_idx['DC_POWER'])) * 100

print("The Conversion Efficiency for Plant 2 is", round(ConvEff_Plant2, 3), "%")

The Conversion Efficiency for Plant 2 is 97.801 %


In [48]:
ConvEff = (sum(ds['AC_POWER']) / (sum(ds['DC_POWER']))) * 100
ConvEff
print("The Conversion Efficiency for the System is", round(ConvEff, 3), "%")

The Conversion Efficiency for the System is 97.798 %


<div style="border-radius:20px;
            border : black solid;
            color: #403b34;
            background-color: #cc7e08;
            font-size:100%;
            text-align: left;
            margin-left: 10px;
            margin-right: 10px;">

<h2 style='; border:0; border-radius: 15px; font-weight: bold; color:#2b2924'><center>
    </br> PERFORMANCE RATIO
</center></h2>  

<ol style="margin-left:10px;margin-left:10px">
    
Represents a metric used to assess the overall performance of a solar panel system and accounts for losses due to shading, soiling, temperature and other factors. It is calculated by dividing the actual energy output (AC power) by the expected energy output based on ideal conditions (irradiation and module temperature) as follows:


</ol>

<ol style="margin-left:360px">
    <b>Performance Ratio = (AC Power / (Irradiation * Module Temperature)) * 100</b>
</ol>


In [49]:
PerformanceRatio = (sum(power_ds['AC_POWER'])/( sum(sensor_ds['IRRADIATION'])*( sum(sensor_ds['MODULE_TEMPERATURE']))))*100

print("The Performance Ratio for the System is", round(PerformanceRatio, 3), "%")

The Performance Ratio for the System is 12.276 %
